codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ValidationGenerator.java

```diff
@@ -166,42 +182,59 @@ private void generateShapeValidationFunctions(
+                        Shape memberTarget = model.expectShape(member.getTarget());
+                        String helperName = getShapeValidationFunctionName(memberTarget, false);
+
                         writer.openBlock("for i := range v {", "}", () -> {
-                            writer.openBlock("if err := $L(v[i]); err != nil {", "}", helperName, () -> {
+                            String addr = "";
+                            if (!pointableIndex.isPointable(member) && pointableIndex.isPointable(memberTarget)) {
+                                addr = "&";
+                            }
+                            writer.openBlock("if err := $L($Lv[i]); err != nil {", "}", helperName, addr, () -> {
                                 writer.addUseImports(SmithyGoDependency.SMITHY);
-                                writer.write(
-                                        "invalidParams.AddNested(fmt.Sprintf(\"[%d]\", i), "
+                                writer.write("invalidParams.AddNested(fmt.Sprintf(\"[%d]\", i), "
                                                 + "err.(smithy.InvalidParamsError))");
                             });
                         });
                         break;
+
                     case MAP:
-                        helperName = getShapeValidationFunctionName(model.expectShape(((MapShape) shape).getValue()
-                                .getTarget()), false);
+                        MapShape mapShape = shape.asMapShape().get();
+                        MemberShape mapValue = mapShape.getValue();
+                        Shape valueTarget = model.expectShape(mapValue.getTarget());
+                        helperName = getShapeValidationFunctionName(valueTarget, false);
+
                         writer.openBlock("for key := range v {", "}", () -> {
-                            writer.openBlock("if err := $L(v[key]); err != nil {", "}", helperName, () -> {
+                            String valueVar = "v[key]";
+                            if (!pointableIndex.isPointable(mapValue) && pointableIndex.isPointable(valueTarget)) {
+                                writer.write("value := $L", valueVar);
+                                valueVar = "&value";
+                            }
+                            writer.openBlock("if err := $L($L); err != nil {", "}", helperName, valueVar, () -> {
                                 writer.addUseImports(SmithyGoDependency.SMITHY);
-                                writer.write(
-                                        "invalidParams.AddNested(fmt.Sprintf(\"[%q]\", key), "
+                                writer.write("invalidParams.AddNested(fmt.Sprintf(\"[%q]\", key), "
                                                 + "err.(smithy.InvalidParamsError))");
                             });
                         });
                         break;
+
                     case UNION:
                         // TODO: Implement Union support
+
```

Comment:
now that we have unions, we should be able to implement support for this. Do we have a tracking task around this validation generator?

Reply:
I don't think there is a tracking task for those, we'll need to add it.

Reply:
Created https://github.com/awslabs/smithy-go/issues/235 for this.